libs/core/langchain_core/graph_stores.py

```diff
@@ -0,0 +1,477 @@
+from __future__ import annotations
+
+from abc import abstractmethod
+from typing import (
+    Any,
+    AsyncIterable,
+    ClassVar,
+    Collection,
+    Iterable,
+    Iterator,
+    List,
+    Optional,
+)
+
+from langchain_core.callbacks import (
+    AsyncCallbackManagerForRetrieverRun,
+    CallbackManagerForRetrieverRun,
+)
+from langchain_core.documents import Document
+from langchain_core.load import Serializable
+from langchain_core.pydantic_v1 import Field
+from langchain_core.runnables import run_in_executor
+from langchain_core.vectorstores import VectorStore, VectorStoreRetriever
+
+
+def _has_next(iterator: Iterator) -> bool:
+    """Checks if the iterator has more elements.
+    Warning: consumes an element from the iterator"""
+    sentinel = object()
+    return next(iterator, sentinel) is not sentinel
+
+
+class Node(Serializable):
+    """Node in the GraphStore."""
+
+    id: Optional[str]
+    """Unique ID for the node. Will be generated by the GraphStore if not set."""
+    metadata: dict = Field(default_factory=dict)
+    """Metadata for the node. May contain information used to link this node
+    with other nodes."""
+
+
+class TextNode(Node):
+    text: str
+    """Text contained by the node."""
+
+
+def _texts_to_nodes(
+    texts: Iterable[str],
+    metadatas: Optional[Iterable[dict]],
+    ids: Optional[Iterable[str]],
+) -> Iterator[Node]:
+    metadatas_it = iter(metadatas) if metadatas else None
+    ids_it = iter(ids) if ids else None
+    for text in texts:
+        try:
+            _metadata = next(metadatas_it) if metadatas_it else {}
+        except StopIteration:
+            raise ValueError("texts iterable longer than metadatas")
+        try:
+            _id = next(ids_it) if ids_it else None
+        except StopIteration:
+            raise ValueError("texts iterable longer than ids")
+        yield TextNode(
+            id=_id,
+            metadata=_metadata,
+            text=text,
+        )
+    if ids_it and _has_next(ids_it):
+        raise ValueError("ids iterable longer than texts")
+    if metadatas_it and _has_next(metadatas_it):
+        raise ValueError("metadatas iterable longer than texts")
+
+
+def _documents_to_nodes(
+    documents: Iterable[Document], ids: Optional[Iterable[str]]
+) -> Iterator[Node]:
+    ids_it = iter(ids) if ids else None
+    for doc in documents:
+        try:
+            _id = next(ids_it) if ids_it else None
+        except StopIteration:
+            raise ValueError("documents iterable longer than ids")
+        yield TextNode(
+            id=_id,
+            metadata=doc.metadata,
+            text=doc.page_content,
+        )
+    if ids_it and _has_next(ids_it):
+        raise ValueError("ids iterable longer than documents")
+
+
+class GraphStore(VectorStore):
```

Comment:
GraphStore sounds purely graph db based. if this is meant to implement hybrid indexing/retrieval methods should we call it sometihng like GraphVectorStore or HybridGraphStore?

and i'd probably put it in langchain_core.vectorstores (can turn the vectorstores.py file into a directory

Reply:
Renamed to GraphVectorStore